lib/Parse/ParseDecl.cpp

```diff
@@ -770,31 +770,72 @@ Parser::parseDifferentiableAttribute(SourceLoc atLoc, SourceLoc loc) {
   ParserStatus Status;
   SourceLoc lParenLoc, rParenLoc;
 
-  // Set parse error, skip until ')' and parse it.
-  auto errorAndSkipToEnd = [&](int parenDepth = 1) -> ParserStatus {
+  // Parse '('.
+  if (!consumeIf(tok::l_paren, lParenLoc)) {
+    diagnose(Tok, diag::attr_expected_lparen, AttrName,
+             /*DeclModifier=*/false);
+    skipUntil(tok::r_paren);
+    if (!consumeIf(tok::r_paren, rParenLoc))
+      diagnose(Tok, diag::attr_expected_rparen, AttrName,
+               /*DeclModifier=*/false);
+    return makeParserError();
+  }
+
+  using FuncSpec = DifferentiableAttr::FunctionSpecifier;
+  AutoDiffMode mode;
+  SourceLoc modeLoc;
```

Comment:
Are `modeLoc`, `primalSpec` and `params` used somewhere after updating inside `parseDifferentiableAttributeArguments`?

Reply:
If I understood you correctly, yes - they're used in `DifferentiableAttr::create`.

Reply:
ah, I see. Could you also add a round-trip test in ./test/Syntax/round_trip_parse_gen.swift to make sure we parse them correctly?

Reply:
Yes, sure thing.
I'm closing this PR and making a new one that points to the `tensorflow` branch instead. I'd appreciate your review on the new PR!